src/java.base/share/classes/java/nio/file/spi/FileSystemProvider.java

```diff
@@ -1170,4 +1170,122 @@ public abstract Map<String,Object> readAttributes(Path path, String attributes,
     public abstract void setAttribute(Path path, String attribute,
                                       Object value, LinkOption... options)
         throws IOException;
+
+    /**
+     * Tests whether a file exists. This method works in exactly the
+     * manner specified by the {@link Files#exists(Path, LinkOption...)} method.
+     *
+     * @implSpec
+     * The default implementation of this method invokes the
+     * {@link #checkAccess(Path, AccessMode...)} method when symbolic links
+     * are followed. If the option {@link LinkOption#NOFOLLOW_LINKS NOFOLLOW_LINKS}
+     * is present then symbolic links are not followed and the method
+     *  {@link #readAttributes(Path, Class, LinkOption...)} is called
+     *  to determine whether a file exists.
+     *
+     * @param   path
+     *          the path to the file to test
+     * @param   options
+     *          options indicating how symbolic links are handled
+     *
+     * @return  {@code true} if the file exists; {@code false} if the file does
+     *          not exist or its existence cannot be determined.
+     *
+     * @throws  SecurityException
+     *          In the case of the default provider, the {@link
+     *          SecurityManager#checkRead(String)} is invoked to check
+     *          read access to the file.
+     *
+     * @since 20
+     */
+    public boolean exists(Path path, LinkOption... options) {
+        try {
```

Comment:
Overall, I think that this is a great change (avoiding the need for various parts of the system to communicate through exceptions).

Trivially, and easily, I often find it useful to add mock-like tests for these specified implementations (implSpec), i.e. just subclass FSP mocking out the abstract methods to ensure that the default implementations invoke the appropriate ones.

Reply:
Thank you for the suggestion Chris